pebbledb.go

```diff
@@ -0,0 +1,271 @@
+	2. Restart new version as normal
+
+
+Example: Upgrading sifchain from v0.14.0 to v0.15.0
+
+# log:
+panic: UPGRADE "0.15.0" NEEDED at height: 8170210: {"binaries":{"linux/amd64":"https://github.com/Sifchain/sifnode/releases/download/v0.15.0/sifnoded-v0.15.0-linux-amd64.zip?checksum=0c03b5846c5a13dcc0d9d3127e4f0cee0aeddcf2165177b2f2e0d60dbcf1a5ea"}}
+
+# step1
+git reset --hard
+git checkout v0.14.0
+go mod edit -replace github.com/tendermint/tm-db=github.com/baabeetaa/tm-db@pebble
+go mod tidy
+go install -tags pebbledb -ldflags "-w -s -X github.com/cosmos/cosmos-sdk/types.DBBackend=pebbledb -X github.com/tendermint/tm-db.PebbleDbForceSync=1" ./cmd/sifnoded
+
+$HOME/go/bin/sifnoded start --db_backend=pebbledb
+
+
+# step 2
+git reset --hard
+git checkout v0.15.0
+go mod edit -replace github.com/tendermint/tm-db=github.com/baabeetaa/tm-db@pebble
+go mod tidy
+go install -tags pebbledb -ldflags "-w -s -X github.com/cosmos/cosmos-sdk/types.DBBackend=pebbledb" ./cmd/sifnoded
+
+$HOME/go/bin/sifnoded start --db_backend=pebbledb
+
+*/
+var PebbleDbForceSync = "0"
+var isPebbleDbForceSync = false
+
+func init() {
+	dbCreator := func(name string, dir string) (DB, error) {
+		return NewPebbleDB(name, dir)
+	}
+
+	registerDBCreator(PebbleDBBackend, dbCreator, false)
+
+	if PebbleDbForceSync == "1" {
+		isPebbleDbForceSync = true
+	} else if PebbleDbForceSync == "0" {
+		isPebbleDbForceSync = false
+	} else {
+		panic("PebbleDbForceSync should be 0 or 1, but provided: " + PebbleDbForceSync)
+	}
+}
+
+type PebbleDB struct {
+	db *pebble.DB
+}
+
+var _ DB = (*PebbleDB)(nil)
+
+func NewPebbleDB(name string, dir string) (*PebbleDB, error) {
+	opts := &pebble.Options{}
+	opts.EnsureDefaults()
+
+	return NewPebbleDBWithOpts(name, dir, opts)
+}
+
+func NewPebbleDBWithOpts(name string, dir string, opts *pebble.Options) (*PebbleDB, error) {
+	dbPath := filepath.Join(dir, name+".db")
+	opts.EnsureDefaults()
+
+	p, err := pebble.Open(dbPath, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &PebbleDB{
+		db: p,
+	}, nil
+}
+
+// Get implements DB.
+func (db *PebbleDB) Get(key []byte) ([]byte, error) {
+	if len(key) == 0 {
+		return nil, errKeyEmpty
+	}
+
+	res, closer, err := db.db.Get(key)
+	if err != nil {
+		if err == pebble.ErrNotFound {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer closer.Close()
+
+	return cp(res), nil
+}
+
+// Has implements DB.
+func (db *PebbleDB) Has(key []byte) (bool, error) {
+	if len(key) == 0 {
+		return false, errKeyEmpty
+	}
+
+	bytes, err := db.Get(key)
+	if err != nil {
+		return false, err
+	}
+
+	return bytes != nil, nil
+}
+
+// Set implements DB.
+func (db *PebbleDB) Set(key []byte, value []byte) error {
+	if len(key) == 0 {
+		return errKeyEmpty
+	}
+
+	if value == nil {
+		return errValueNil
+	}
+
+	wopts := pebble.NoSync
+	if isPebbleDbForceSync {
+		wopts = pebble.Sync
+	}
```

Comment:
I think it should be possible to remove global var lookup here by adding a wrapper for the pebbledb

e.g.

```
type PebbleDBWrapper struct {
    ...
}

func (db *PebbleDBWrapper) isForceSync() bool { ... }
```

Reply:
Decided not to